### PR TITLE
Mock array element latency

### PR DESCRIPTION
**Pull request by tspyrou**

### CODE REVIEW

**Comment by tspyrou** [comment]
New PR with the metrics code enhanced to capture the clock latency for both the source and the target clocks. From this we can make the element constraints.sdc file appear to have close to the same slack as the flattened version.

I deleted the old PR with the copious notes to self.

**Review by oharboe — CHANGES_REQUESTED** [review]
I need a sample mock-array/Element constraints.sdc file that shows how to use the information in this report to make meaningful comments.

**Comment by tspyrou** [comment]
@oharboe I will add one.

**Comment by tspyrou** [comment]
@oharboe I updated the block SDC. The numbers in it come from analyzing this information in the updated metrics report.
The goal is to make the path inside the element look like the path at the top.
We could change the input delay in the element sdc to match the 50 seen at the top but I did not modify that heuristic.
This is from the flat run.
<pre>
 ==========================================================================
global route post repair timing report_checks -path_delay min from inputs
--------------------------------------------------------------------------
Startpoint: io_ins_left_3[0] (input port clocked by clock_vir)
Endpoint: ces_2_7.io_ins_down[0]$_DFF_P_
          (rising edge-triggered flip-flop clocked by clock)
Path Group: clock
Path Type: min

  Delay    Time   Description
---------------------------------------------------------
   0.00    0.00   clock clock_vir (rise edge)
 380.00  380.00   clock network delay (ideal)
  50.00  430.00 v input external delay
   0.00  430.00 v io_ins_left_3[0] (in)
  11.06  441.06 v input705/Y (BUFx2_ASAP7_75t_R)
   0.10  441.16 v ces_2_7.io_ins_down[0]$_DFF_P_/D (DFFHQNx3_ASAP7_75t_R)
         441.16   data arrival time

   0.00    0.00   clock clock (rise edge)
 397.76  397.76   clock network delay (propagated)
  10.00  407.76   clock uncertainty
   0.00  407.76   clock reconvergence pessimism
         407.76 ^ ces_2_7.io_ins_down[0]$_DFF_P_/CLK (DFFHQNx3_ASAP7_75t_R)
  13.52  421.28   library hold time
         421.28   data required time
---------------------------------------------------------
         421.28   data required time
        -441.16   data arrival time
---------------------------------------------------------
          19.88   slack (MET)



==========================================================================
global route post repair timing critical path target clock latency max path
--------------------------------------------------------------------------
310.4272

==========================================================================
global route post repair timing critical path target clock latency min path
--------------------------------------------------------------------------
397.7622

==========================================================================
global route post repair timing critical path source clock latency min path
--------------------------------------------------------------------------
380.0000
</pre>

**Comment by tspyrou** [comment]
This is what the path looks like at the Element level with the new SDC for hold.

<pre>
==========================================================================
cts final setup_violation_count
--------------------------------------------------------------------------
setup violation count 264

==========================================================================
cts final hold_violation_count
--------------------------------------------------------------------------
hold violation count 0

==========================================================================
cts final report_checks -path_delay min from inputs
--------------------------------------------------------------------------
Startpoint: io_ins_left[25] (input port clocked by clock_vir)
Endpoint: io_outs_down[25]$_DFF_P_
          (rising edge-triggered flip-flop clocked by clock)
Path Group: clock
Path Type: min

  Delay    Time   Description
---------------------------------------------------------
   0.00    0.00   clock clock_vir (rise edge)
 450.00  450.00   clock network delay (ideal)
  30.00  480.00 v input external delay
   0.70  480.70 v io_ins_left[25] (in)
  14.51  495.22 v input82/Y (BUFx2_ASAP7_75t_R)
   0.29  495.51 v io_outs_down[25]$_DFF_P_/D (DFFHQNx2_ASAP7_75t_R)
         495.51   data arrival time

   0.00    0.00   clock clock (rise edge)
 466.31  466.31   clock network delay (propagated)
  20.00  486.31   clock uncertainty
   0.00  486.31   clock reconvergence pessimism
         486.31 ^ io_outs_down[25]$_DFF_P_/CLK (DFFHQNx2_ASAP7_75t_R)
   9.18  495.49   library hold time
         495.49   data required time
---------------------------------------------------------
         495.49   data required time
        -495.51   data arrival time
---------------------------------------------------------
           0.02   slack (MET)

</pre>

This is the top level path
<pre>
==========================================================================
cts post-repair setup_violation_count
--------------------------------------------------------------------------
setup violation count 2112

==========================================================================
cts post-repair hold_violation_count
--------------------------------------------------------------------------
hold violation count 2048
==========================================================================
cts post-repair report_checks -path_delay min from inputs
--------------------------------------------------------------------------
Startpoint: io_ins_right_0[59] (input port clocked by clock_vir)
Endpoint: ces_0_0 (rising edge-triggered flip-flop clocked by clock)
Path Group: clock
Path Type: min

  Delay    Time   Description
---------------------------------------------------------
   0.00    0.00   clock clock_vir (rise edge)
   0.00    0.00   clock network delay (ideal)
  50.00   50.00 v input external delay
   0.00   50.00 v io_ins_right_0[59] (in)
  11.48   61.48 v input1079/Y (BUFx2_ASAP7_75t_R)
   0.07   61.55 v ces_0_0/io_ins_right[59] (Element)
          61.55   data arrival time

   0.00    0.00   clock clock (rise edge)
   0.00    0.00   clock source latency
   0.00    0.00 ^ clock (in)
  19.58   19.58 ^ wire2/Y (BUFx16f_ASAP7_75t_R)
  65.14   84.72 ^ wire1/Y (BUFx16f_ASAP7_75t_R)
  98.92  183.63 ^ clkbuf_0_clock/Y (BUFx24_ASAP7_75t_R)
  32.55  216.19 ^ clkbuf_1_0_0_clock/Y (BUFx24_ASAP7_75t_R)
  35.68  251.87 ^ clkbuf_2_0_0_clock/Y (BUFx24_ASAP7_75t_R)
  34.70  286.56 ^ clkbuf_3_0_0_clock/Y (BUFx24_ASAP7_75t_R)
  67.90  354.46 ^ ces_0_0/clock (Element)
  10.00  364.46   clock uncertainty
   0.00  364.46   clock reconvergence pessimism
  64.98  429.45   library hold time
         429.45   data required time
---------------------------------------------------------
         429.45   data required time
         -61.55   data arrival time
---------------------------------------------------------
        -367.89   slack (VIOLATED)


</pre>

**Comment by tspyrou** [comment]
@oharboe My last comment about the scripts and constraints.sdc was in error. I deleted it.
write_timing_model is erroneously including the clock source latency into the model which shows up in a massive required hold time. This is one reason why the run fails unable to fix hold.

**Comment by tspyrou** [comment]
@oharboe sorry for all of the comments. I also see that the flat run is not a good predictor of the hierarchical run for some reason.   I had used the flat run to get an idea of the latencies. I will try that with the hier run instead.

**Comment by tspyrou** [comment]
@oharboe I think we can reserve some time when you are here to discuss.
Currently the register to register timing is closing at the top level and it gets stuck trying to close the i/o hold violations.
Its good to separate those variables and the report_metrics here does that.

File: flow/scripts/generate_abstract.tcl [hunk 1]
```diff
@@ -16,7 +16,8 @@ if {$design_stage >= 6 && [file exists $::env(RESULTS_DIR)/$stem.spef]} {
 if {$design_stage >= 4} {
   set_propagated_clock [all_clocks]
 }
-
+# write_timing_model includes the source latency in the model
+set_clock_latency -source 0 [all_clocks]
```
Comment (maliberty) [inline flow/scripts/generate_abstract.tcl:20]: Should sta handle this?  This seems error prone

**Comment by maliberty** [comment]
Is this generally useful?  I'm not clear what users should do with this information.

**Pull request merged**